04-onion-routing.md

```diff
@@ -376,6 +413,180 @@ otherwise meets the amount criterion (eg. some other failure, or
 invoice timeout), however if it were to fulfill only some of them,
 intermediary nodes could simply claim the remaining ones.
 
+### Route Blinding
+
+Nodes receiving onion packets may hide their identity from senders by
+"blinding" an arbitrary amount of hops at the end of an onion path.
+
+When using route blinding, nodes find a route to themselves from a given
+"introduction node". They then use ECDH with each node in that route to create
+a "blinded" node ID and an encrypted blob (`encrypted_data`) for each one of
+the blinded nodes.
+
+They communicate this blinded route and the encrypted blobs to the sender.
+The sender finds a route to the introduction node and extends it with the
+blinded route provided by the recipient. The sender includes the encrypted
+blobs in the corresponding onion payloads: they allow nodes in the blinded
+part of the route to "unblind" the next node and correctly forward the packet.
+
+The `encrypted_data` is a TLV stream, encrypted for a given blinded node, that
+may contain the following TLV fields:
+
+1. `tlv_stream`: `encrypted_data_tlv`
+2. types:
+    1. type: 1 (`padding`)
+    2. data:
+        * [`...*byte`:`padding`]
+    1. type: 2 (`short_channel_id`)
+    2. data:
+        * [`short_channel_id`:`short_channel_id`]
+    1. type: 4 (`next_node_id`)
+    2. data:
+        * [`point`:`node_id`]
+    1. type: 6 (`path_id`)
+    2. data:
+        * [`...*byte`:`data`]
+    1. type: 8 (`next_blinding_override`)
+    2. data:
+        * [`point`:`blinding`]
+    1. type: 10 (`payment_relay`)
+    2. data:
+        * [`u32`:`fee_base_msat`]
+        * [`u32`:`fee_proportional_millionths`]
+        * [`u16`:`cltv_expiry_delta`]
+    1. type: 12 (`payment_constraints`)
+    2. data:
+        * [`u32`:`max_cltv_expiry`]
+        * [`u64`:`htlc_minimum_msat`]
+        * [`...*byte`:`allowed_features`]
```

Comment:
Why do we need that? What features would you put here?
We should let the relaying node do whatever it wants as long as it relays the right amount.

Reply:
There is a paragraph explaining that in the proposals doc: https://github.com/lightning/bolts/blob/route-blinding/proposals/route-blinding.md#blinded-payments

Imagine for example that we add support for PTLCs, and 50% of the network supports it. If we didn't have this requirement, an attacker could make a PTLC payment over the blinded route. If it works, that reduces the anonymity set by 50%. Privacy-aware receivers would make sure that they don't include this feature bit to avoid reducing their anonymity set (they would start including this feature bit only once most of the network would support the feature and it wouldn't be a fingerprinting vector anymore). Does that make sense? Basically anything that can let the attacker reduce the anonymity set must be decided by the recipient and included in `encrypted_data`.

Right now the only feature bit that could be used as a fingerprint for relay is `var_onion_optin`, but since the whole network supports it we can probably omit it and assume support anyway, so this field will be omitted for now.

Reply:
OK, then the set of features that need to be included to be allowed must be explicitly listed.

Reply:
Agreed, it's worth adding to the table in Bolt 9. But I'm waiting for feedback from other implementations before adding this, I'd like to see if there is rough consensus first.

Reply:
Added in https://github.com/lightning/bolts/pull/765/commits/83c44dc7476e4f5ec85c531eca09c617cd7f97a4